sup/concurrency.cpp

```diff
@@ -52,26 +57,39 @@ util::optional<size_t> get_env_num_threads() {
     if (errno==ERANGE ||
         !std::regex_match(str, std::regex("\\s*\\d*[0-9]\\d*\\s*")))
     {
-        throw arbor_exception(util::pprintf(
-            "requested number of threads \"{}\" is not a valid value", str));
+        throw arb::arbor_exception(
+            std::string("requested number of threads \"") + str + "\" is not a valid value");
     }
+    errno = 0;
 
     return nthreads;
 }
 
-std::size_t num_threads_init() {
-    std::size_t n = 0;
+// Take a best guess at the number of threads that can be run concurrently.
+// Will return at least 1.
+unsigned thread_concurrency() {
+    // Attempt to get count first from affinity information if available.
+    unsigned n = get_affinity().size();
 
-    if (auto env_threads = get_env_num_threads()) {
-        n = env_threads.value();
+    // If no luck, try sysconf.
+#ifdef _SC_NPROCESSORS_ONLN
+    if (!n) {
+        long r = sysconf(_SC_NPROCESSORS_ONLN);
+        if (r>0) {
+            n = (unsigned)r;
+        }
     }
+#endif
 
+    // If still zero, try the hint from the library.
     if (!n) {
-        n = hw::node_processors();
+        n = std::thread::hardware_concurrency();
     }
 
-    return n? n: 1;
+    // If still zero, use one thread.
+    n = std::max(n, 1u);
+
+    return n;
```

Comment:
We can keep the comment, but using `max` here just adds code to no effect: there's nothing wrong with `n? n: 1`.

Reply:
I personally prefer using `std::max` for this sort of logic. It is arguable whether a ternary or `max` is clearer... probably a matter of what you are used to.
The suggested ternary `n? n: 1` is short hand for `n>0? n: 1`, so while it is less "code", it isn't less logic for the reader to mentally parse.

That is my five cents, though I have no problem making the change.

Reply:
For me, both have equal cognitive load, but I'm less tempted to think about possible c++ weirdness with `n ? n : 1` than `max`. (And their not precisely equivalent but I have no idea what the compiler could do to either one).

Reply:
The compiler should generate equally efficient code, it is just a matter of semantics I think.

https://godbolt.org/z/0ITuiT

Reply:
Man, I can never convince gcc to produce a conditional move when I want it to, damn it.